File: DesingPattern/AbstractFactory/AbstractFactory.js
class AnimalFactory {
    createAnimal() {
        throw new Error('This method should be overrriden');
    }
}

class DogFactory extends AnimalFactory {
    createAnimal() {
        return new Dog();
    }
}

class CatFactory extends AnimalFactory {
    createAnimal() {
        return new Cat();
    }
}

class Animal {
   makeSound() {
    throw new Error('This method should be overridden');
    }
}


class Dog extends Animal {
  makeSound() {
    return "Woof!";
  }
}

// Concrete Product 2
class Cat extends Animal {
  makeSound() {
    return "Meow!";
  }
}

// Usage 
const dogFactory = new DogFactory();
const dog = dogFactory.createAnimal();
console.log(dog.makeSound()); // Output: "Woof!"

const catFactory = new CatFactory();
const cat = catFactory.createAnimal();
console.log(cat.makeSound()); // Output: "

File: DesingPattern/AbstractFactory/AbstractFactory.py
# Abstract Factory
class AnimalFactory:
    def create_animal(self):
        raise NotImplementedError("This method should be overridden")


# Concrete Factory 1
class DogFactory(AnimalFactory):
    def create_animal(self):
        return Dog()


# Concrete Factory 2
class CatFactory(AnimalFactory):
    def create_animal(self):
        return Cat()


# Abstract Product
class Animal:
    def make_sound(self):
        raise NotImplementedError("This method should be overridden")


# Concrete Product 1
class Dog(Animal):
    def make_sound(self):
        return "Woof!"


# Concrete Product 2
class Cat(Animal):
    def make_sound(self):
        return "Meow!"


# Usage
dog_factory = DogFactory()
dog = dog_factory.create_animal()
print(dog.make_sound())  # Output: "Woof!"

cat_factory = CatFactory()
cat = cat_factory.create_animal()
print(cat.make_sound())  # Output: "Meow!"
